linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -191,9 +194,16 @@ final class Emitter(config: Emitter.Config) {
     val moduleTrees = logger.time("Emitter: Write trees") {
       moduleSet.modules.map { module =>
         val moduleContext = ModuleContext.fromModule(module)
+        val moduleCache = state.moduleCaches.getOrElseUpdate(module.id, new ModuleCache)
 
         val moduleClasses = generatedClasses(module.id)
 
+        val moduleImports = extractWithGlobals {
+          moduleCache.getOrComputeImports(module.externalDependencies, module.internalDependencies) {
+            genModuleImports(module)
+          }
```

Comment:
I'm not a fan of the separate dependency / generation tracking here (i.e. it isn't obvious that we're guarding on all the necessary things).

However, this is the general pattern in the emitter right now, so we should probably address it later.